doc/sphinx/proof-engine/tactics.rst

```diff
@@ -898,211 +1109,156 @@ Applying theorems
+   We recommend always specifying :n:`@ident` so that the names of hypotheses don't
+   change as the proof is updated, making your proof easier to maintain.  For example,
+   if H exists in the context, Coq will consider using `H0`, `H1`, ... until it finds an
+   unused name.  Modifications to a proof can change automatically assigned names
+   that subsequent tactics likely refer to, making the proofs harder to maintain.
 
-:opt:`Debug` ``"tactic-unification"`` enables printing traces of
-unification steps in tactic unification. Tactic unification is used in
-tactics such as :tacn:`apply` and :tacn:`rewrite`.
+   Note that :tacn:`intros` lets you introduce multiple items into
+   the context with a single tactic.
 
-.. _managingthelocalcontext:
+   :n:`@ident`
+     The name to give to the introduced hypothesis.  If not given, Coq uses the
+     variable name for dependent premises or `H` for non-dependent premises.
+     If a name such as `H` is already in use, Coq will consider using `H0`,
+     `H1`, ... until it finds a fresh name.
 
-Managing the local context
-------------------------------
+     .. note::
 
-.. tacn:: intro
-   :name: intro
+        If a hypothesis name hides the base name of a global constant then
+        the latter can still be referred to by a qualified name
+        (see :ref:`Qualified-names`).
 
-   This tactic applies to a goal that is either a product or starts with a
-   let-binder. If the goal is a product, the tactic implements the "Lam" rule
-   given in :ref:`Typing-rules` [1]_. If the goal starts with a let-binder,
-   then the tactic implements a mix of the "Let" and "Conv".
+   :n:`@where`
+     Indicates where to place the introduced hypothesis: at the top or bottom
+     of the context or before or after another specified hypothesis.  The default
+     is `at bottom`.
 
-   If the current goal is a dependent product :g:`forall x:T, U`
-   (resp :g:`let x:=t in U`) then :tacn:`intro` puts :g:`x:T` (resp :g:`x:=t`)
-   in the local context. The new subgoal is :g:`U`.
-
-   If the goal is a non-dependent product :math:`T \rightarrow U`, then it
-   puts in the local context either :g:`Hn:T` (if :g:`T` is of type :g:`Set`
-   or :g:`Prop`) or :g:`Xn:T` (if the type of :g:`T` is :g:`Type`).
-   The optional index ``n`` is such that ``Hn`` or ``Xn`` is a fresh
-   identifier. In both cases, the new subgoal is :g:`U`.
-
-   If the goal is an existential variable, :tacn:`intro` forces the resolution
-   of the existential variable into a dependent product :math:`\forall`\ :g:`x:?X, ?Y`,
-   puts :g:`x:?X` in the local context and leaves :g:`?Y` as a new subgoal
-   allowed to depend on :g:`x`.
+   .. exn:: @ident is already used.
+      :undocumented:
 
-   The tactic :tacn:`intro` applies the tactic :tacn:`hnf`
-   until :tacn:`intro` can be applied or the goal is not head-reducible.
+   .. exn:: No such hypothesis: @ident.
+      :undocumented:
 
    .. exn:: No product even after head-reduction.
```

Comment:
Raised when there is nothing to introduce even after `hnf` has been completely applied.